client_new/src/types/client.rs

```diff
@@ -0,0 +1,207 @@
+
+        ClientVault {
+            client: self.atomic_ref(),
+            id: vault_id,
+        }
+    }
+
+    /// Returns `true`, if a vault exists
+    ///
+    /// # Example
+    /// ```
+    /// ```
+    pub async fn vault_exists<P>(&self, vault_path: P) -> Result<bool, ClientError>
+    where
+        P: AsRef<Vec<u8>>,
+    {
+        let vault_id = derive_vault_id(vault_path);
+        let keystore = self.keystore.try_read().map_err(|_| ClientError::LockAcquireFailed)?;
+
+        Ok(keystore.vault_exists(vault_id))
+    }
+
+    /// Returns Ok, if the record exists
+    ///
+    /// # Example
+    /// ```
+    /// ```
+    pub async fn record_exists(&mut self, location: Location) -> Result<bool, ClientError> {
+        let (vault_id, record_id) = location.resolve();
+        let db = self.db.try_read().map_err(|_| ClientError::LockAcquireFailed)?;
+        let contains_record = db.contains_record(vault_id, record_id);
+        Ok(contains_record)
+    }
+
+    /// Returns the [`ClientId`] of the client
+    ///
+    /// # Example
+    /// ```
+    /// ```
+    pub async fn id(&self) -> &ClientId {
+        &self.id
+    }
+
+    /// Loads the state of [`Self`] from a [`ClientState`]. Replaces all previous data.
+    ///
+    /// # Example
+    /// ```
+    /// ```
+    pub async fn load(&self, state: ClientState, id: ClientId) -> Result<(), ClientError> {
+        let (keys, db, st) = state;
+
+        // reload keystore
+        let mut keystore = self.keystore.try_write().map_err(|_| ClientError::LockAcquireFailed)?;
+        let mut new_keystore = KeyStore::<Provider>::default();
+        new_keystore
+            .rebuild_keystore(keys)
+            .map_err(|e| ClientError::Inner(e.to_string()))?;
+
+        *keystore = new_keystore;
+        drop(keystore);
+
+        // reload db
+        let mut view = self.db.try_write().map_err(|_| ClientError::LockAcquireFailed)?;
+        *view = db;
+        drop(view);
+
+        // reload store
+        let mut store = self
+            .store
+            .cache
+            .try_write()
+            .map_err(|_| ClientError::LockAcquireFailed)?;
+        *store = st;
+        drop(store);
+
+        Ok(())
+    }
+
+    /// Executes a cryptographic [`Procedure`] and returns its output.
+    /// A cryptographic [`Procedure`] is the main operation on secrets.
+    ///
+    /// # Example
+    /// ```no_run
+    /// ```
+    pub async fn execute_procedure<P>(&self, procedure: P) -> Result<P::Output, ProcedureError>
+    where
+        P: Procedure + Into<StrongholdProcedure>,
+    {
+        let res = self.execure_procedure_chained(vec![procedure.into()]).await;
+        let mapped = res.map(|mut vec| vec.pop().unwrap().try_into().ok().unwrap())?;
+        Ok(mapped)
+    }
+
+    /// Executes a list of cryptographic [`Procedures`] sequentially and returns a collected output
+    ///
+    /// # Example
+    /// ```no_run
+    /// ```
+    pub async fn execure_procedure_chained(
```

Comment:
Why are some of the methods on the `Client` async? Unless I am missing something, there is no need for it.

Reply:
true, it's a reminiscent of the old interface. Most of the operations are sync, since we do not enforce a hard dependency on some async runtime. I/O operations on the `Snapshot` should be async, imho. I wonder, if we should enforce a dependency on tokio (or any other async runtime)